control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+      } else if (externFunction->method->name == "resubmit") {
+        auto name = externalNameFromTypeArgs(call->typeArguments, refMap, "resubmit");
+        serializeFieldListToPI(serializer, typeMap, name, call->arguments->at(0));
+      } else if (externFunction->method->name == "recirculate") {
+        auto name = externalNameFromTypeArgs(call->typeArguments, refMap, "recirculate");
+        serializeFieldListToPI(serializer, typeMap, name, call->arguments->at(0));
+      }
+    });
+  };
+}
+
+static void collectExternSymbols(PISymbolTable& symbols,
+                                 const IR::ExternBlock* externBlock) {
+  CHECK_NULL(externBlock);
+
+  if (externBlock->type->name == "counter") {
+    symbols.add(PISymbolType::COUNTER, externBlock->node->to<IR::IDeclaration>());
+  } else if (externBlock->type->name == "meter") {
+    symbols.add(PISymbolType::METER, externBlock->node->to<IR::IDeclaration>());
+  }
+}
+
+static void serializeExternToPI(P4InfoSerializer& serializer,
+                                const IR::ExternBlock* externBlock,
+                                const TypeMap* typeMap) {
+  CHECK_NULL(externBlock);
+  CHECK_NULL(typeMap);
+
+  if (externBlock->type->name == "counter") {
+    serializer.addCounter(externBlock->node->to<IR::IDeclaration>(),
+                          externBlock->getParameterValue("size"),
+                          externBlock->getParameterValue("type"));
+  } else if (externBlock->type->name == "meter") {
+    serializer.addMeter(externBlock->node->to<IR::IDeclaration>(),
+                        externBlock->getParameterValue("size"),
+                        externBlock->getParameterValue("type"));
+  }
+}
+
+static void collectParserSymbols(PISymbolTable& symbols,
+                                 const IR::ParserBlock* parserBlock,
+                                 TypeMap* typeMap) {
+  CHECK_NULL(parserBlock);
+  CHECK_NULL(parserBlock->container);
+  CHECK_NULL(typeMap);
+
+  auto parserType = parserBlock->container->type;
+  for (auto param : *parserType->applyParams->getEnumerator()) {
+    if (param->direction != IR::Direction::Out &&
+        param->direction != IR::Direction::InOut) {
+      // This parameter isn't an output, so nothing downstream can match on it.
+      // That means that we don't need to expose it in PI.
+      continue;
+    }
+
+      // This parameter is an output; it may contain header fields.
+    forAllHeaderFields(param, typeMap, [&](const HeaderField* field) {
+      symbols.addHeaderField(field);
+    });
+  }
+}
+
+static void serializeParserToPI(P4InfoSerializer& serializer,
+                                const IR::ParserBlock* parserBlock,
+                                const TypeMap* typeMap) {
+  CHECK_NULL(parserBlock);
+  CHECK_NULL(parserBlock->container);
+  CHECK_NULL(typeMap);
+
+  auto parserType = parserBlock->container->type;
+  for (auto param : *parserType->applyParams->getEnumerator()) {
```

Comment:
This just mirrors what the BMV2 JSON generator does. It makes sense to me to some extent (after all, parsers are the things which populate headers), but I agree that it seems like we should be able to handle this more cleanly. I just need to learn more to understand what the best approach would be.